pkg/chunked/storage_linux.go

```diff
@@ -1609,6 +1615,20 @@ func (c *chunkedDiffer) ApplyDiff(dest string, options *archive.TarOptions, diff
 		}
 	}
 
+	if output.UncompressedDigest == "" && output.TarSplit != nil {
+		metadata := tsStorage.NewJSONUnpacker(bytes.NewReader(output.TarSplit))
+		fg, err := newStagedFileGetter(dirfd, differOpts, entriesWithOriginalNames)
+		if err != nil {
+			return output, fmt.Errorf("preparing stagedFileGetter: %w", err)
+		}
+		defer fg.Close()
+		digester := digest.Canonical.Digester()
+		if err := asm.WriteOutputTarStream(fg, metadata, digester.Hash()); err != nil {
```

Comment:
This is the expensive part where we're basically re-synthesizing the tarball from the unpacked files and the tar split-data just to get the checksum, right?

I know this is just a PoC but it seems very much worth a comment at the top of this conditional, something like:

```
// If we don't yet have the uncompressed digest, compute it now. This is needed
// for zstd:chunked for example to ensure callers identifying images by this
// have the data.
```

Reply:
> This is the expensive part where we're basically re-synthesizing the tarball from the unpacked files and the tar split-data just to get the checksum, right?

Yes.

Reply:
Comment added, discussing more of the goals, costs and tradeoffs.